curator-framework/src/main/java/org/apache/curator/framework/imps/CuratorFrameworkImpl.java

```diff
@@ -430,6 +448,8 @@ public void close()
             {
                 ensembleTracker.close();
             }
+            backgroundOperations.forEach(this::closeOperation);
+            backgroundOperations.clear();
```

Comment:
Perhaps following these line https://github.com/apache/curator/pull/464/files#diff-1c44d20ca65904e08924d5b0fbc2e3e34d2a043f71f14aa1386e9f45a95ada52R1105-R1107

... to drain the queue first and then process ops one by one. I guess BlockingQueue.foreach is somehow a smell.

Reply:
Done. Added a fixup commit for this and correct a direct call to `backgroundOperations.offer(operationAndData)`.

Reply:
Hmm, the `DelayQueue::drainTo` filters elements based on `Delayed::getDelay`. This is not what we want here. The test could not last 200 runs. I plan to resort to `toArray` for this.